Reference Space Warp Transformation for ZVZB Forces

rcclay opened the pull request:
## Proposed changes

This implements a preliminary reference version of the space-warp transformation for ZVZB force estimators, given in https://doi.org/10.1063/1.3516208.  Currently, Grad_i(E_L) w.r.t. electron coordinates is handled through finite differences, but replacement with analytic derivatives should reasonably occur within the current infrastructure.  

## What type(s) of changes does this code introduce?

- New feature


### Does this introduce a breaking change?

- No

## What systems has this change been tested on?
Local intel Xeon server.
## Checklist

- Yes. This PR is up to date with current the current state of 'develop'
- No. Code added or changed in the PR has been clang-formatted
- Yes. This PR adds tests to cover any new code, or to catch a bug that is being fixed.  
- No. Documentation has been added (if appropriate)

ye-luo commented on src/QMCHamiltonians/ACForce.cpp:
```diff
@@ -15,18 +15,23 @@
 {
   prefix = "ACForce";
   myName = prefix;
   Nions  = ions.getTotalNum();
```
please make Nions const in its definition and assign its value at the constructor initializer list.

ye-luo commented on src/QMCHamiltonians/ACForce.cpp:
```diff
@@ -52,10 +78,20 @@ ACForce::Return_t ACForce::evaluate(ParticleSet& P)
+  
+  if(useSpaceWarp)
+  {
+    ham.evaluateElecGrad(P,psi,el_grad,delta);
```
move the el_grad inside if(useSpaceWarp)

ye-luo commented on src/QMCHamiltonians/ACForce.cpp:
```diff
@@ -75,10 +111,25 @@ void ACForce::addObservables(PropertySetType& plist, BufferType& collectables)
       plist.add(wfgradname1.str());
       plist.add(wfgradname2.str());
+
+      //Remove when ACForce is production ready.
```
Add `TODO: ` before Remove.

ye-luo commented on src/QMCHamiltonians/ACForce.h:
```diff
@@ -54,6 +55,8 @@ struct ACForce : public OperatorBase
   /** Evaluate **/
   Return_t evaluate(ParticleSet& P);
 
+  RealType delta; //finite difference time step
```
Please use Doxygen style comment

ye-luo commented on src/QMCHamiltonians/OperatorBase.h:
```diff
@@ -246,6 +246,7 @@ struct OperatorBase : public QMCTraits
    *@return the value of the Hamiltonian component
    */
   virtual Return_t evaluate(ParticleSet& P) = 0;
+  virtual Return_t evaluateDeterministic(ParticleSet& P);
```
Documentation is needed for this function. Please describe the desired behaviour.

ye-luo replied:
Could you say a bit more about waht "deterministically" means?

ye-luo commented on src/QMCHamiltonians/OperatorBase.cpp:
```diff
@@ -35,6 +35,10 @@ OperatorBase::OperatorBase() : myIndex(-1), Dependants(0), Value(0.0), tWalker(0
 
+OperatorBase::Return_t OperatorBase::evaluateDeterministic(ParticleSet& P)
+{
+  return evaluate(P);
```
Is this a safe fallback? The fallback seems non-deterministic.

rcclay replied:
Unfortunately, the correct behavior of this routine requires estimators with nondeterministic components in their evaluate() have this method overloaded.  Provided that is done, it's safe.  Is this what you were referring to?

ye-luo replied:
"the correct behavior of this routine requires estimators with nondeterministic components in their evaluate() have this method overloaded" this piece of info should be recorded in the OperatorBase.h before the function.

ye-luo commented on src/QMCHamiltonians/SpaceWarpTransformation.h:
```diff
@@ -0,0 +1,62 @@
+  ~SpaceWarpTransformation(){};
+
+ 
+  RealType f(RealType r);  //space warp transformation function F.
```
please use doxygen style comments for functions.

ye-luo commented on src/QMCHamiltonians/SpaceWarpTransformation.h:
```diff
@@ -0,0 +1,62 @@
+                                                                                                                        //this to compute the ZV (el_contribution) and ZB (psi_contribution) 
+                                                                                                                        //space warp contributions.                                                               
+
+  int myTableIndex;
```
const myTableIndex, Nelec, Nions

ye-luo commented on src/QMCHamiltonians/SpaceWarpTransformation.cpp:
```diff
@@ -0,0 +1,102 @@
+namespace qmcplusplus
+{
+
+SpaceWarpTransformation::SpaceWarpTransformation(ParticleSet& elns, ParticleSet& ions):
```
`const ParticleSet& ions`

ye-luo commented on src/QMCHamiltonians/ACForce.cpp:
```diff
@@ -15,18 +15,23 @@
 {
 ACForce::ACForce(ParticleSet& source, ParticleSet& target, TrialWaveFunction& psi_in, QMCHamiltonian& H)
-    : ions(source), elns(target), psi(psi_in), ham(H), FirstForceIndex(-1), Nions(0)
+    : ions(source), elns(target), psi(psi_in), ham(H), FirstForceIndex(-1), Nions(0), useSpaceWarp(false), swt(target,source)
```
Does ACForce modify ions? If no, it is better to set `ions` and `source` as const reference.

rcclay replied:
ACForce potentially modifies ions, if QMCHamiltonian.evaluateIonDerivs starts doing finite differences or something.  Would be complicated to make this const here.  For SpaceWarpTransformation, there's no good reason the ions should be modified, so I'll make those consts.

ye-luo replied:
That makes sense.

ye-luo commented on src/QMCHamiltonians/SpaceWarpTransformation.h:
```diff
@@ -0,0 +1,62 @@
+ *       J. Chem. Phys., 133, 23411 (2010), https://doi.org/10.1063/1.3516208 
+ *
+ */
+struct SpaceWarpTransformation : public QMCTraits
```
please use class instead of struct and put proper public, private.

ye-luo reviewed (changes requested):
Please clang-format the modified source code will also be helpful.

rcclay commented:
@ye-luo  OK.  Short of the making ions "const" issue and the default behavior of evaluateDeterministic, I believe I have addressed all your concerns.

ye-luo commented on src/QMCHamiltonians/NonLocalECPotential.h:
```diff
@@ -152,7 +152,7 @@ class NonLocalECPotential : public OperatorBase, public ForceBase
    * @param Tmove whether Txy for Tmove is updated
    */
-  void evaluateImpl(ParticleSet& P, bool Tmove);
+  void evaluateImpl(ParticleSet& P, bool Tmove, bool keepGrid=false);
```
please document keepGrid. This file seems not being formatted.

ye-luo commented on src/QMCHamiltonians/QMCHamiltonian.h:
```diff
@@ -231,6 +231,12 @@ class QMCHamiltonian
+  /** evaluate Local Energy deterministically.  Defaults to evaluate(P) for operators without a stochastic component. For the nonlocal PP, the quadrature grid is not rerandomized.  
+   * @param P ParticleSet
+   * @return Local energy. 
+   *
```
Could you remove this additional empty line.

ye-luo commented on src/QMCHamiltonians/SpaceWarpTransformation.h:
```diff
@@ -0,0 +1,99 @@
+  public:
+
+  SpaceWarpTransformation(ParticleSet& elns, const ParticleSet& ions);
+  ~SpaceWarpTransformation(){};
```
Please remove this empty destructor line.

ye-luo commented on src/QMCHamiltonians/SpaceWarpTransformation.h:
```diff
@@ -0,0 +1,99 @@
+  ~SpaceWarpTransformation(){};
+
+  RealType f(RealType r);  ///< Space warp transformation function F.  Argument is distance.
+  RealType df(RealType r); ///< The derivative of F w.r.t. r.
```
Please use doxygen style.

ye-luo commented on src/QMCHamiltonians/SpaceWarpTransformation.h:
```diff
@@ -0,0 +1,99 @@
+                                                         
+
+
+  const int myTableIndex;
```
please comment what table it is? My guess is electron ion table.

ye-luo reviewed (commented):
Thank you for helping with the code quality improvements. Only a few minor issues left.
